src/thumbnailer.cc

```diff
@@ -137,31 +110,55 @@ Thumbnailer::Status Thumbnailer::GenerateAnimation(WebPData* const webp_data) {
     }
   }
 
+  config_.quality = final_quality;
+
   if (final_quality == -1) {
     return kByteBudgetError;
   }
 
-  config_.quality = final_quality;
-
   return kOk;
 }
 
+int Thumbnailer::GetPSNR(WebPPicture* const pic, int quality) {
+  WebPPicture new_pic;
+  WebPPictureInit(&new_pic);
+  WebPPictureCopy(pic, &new_pic);
+
+  WebPAuxStats stats;
+  new_pic.stats = &stats;
+
+  config_.quality = quality;
+  WebPEncode(&config_, &new_pic);
+  int result_psnr = new_pic.stats->PSNR[3];  // PSNR-all.
+  WebPPictureFree(&new_pic);
+
+  return result_psnr;
+}
+
 Thumbnailer::Status Thumbnailer::GenerateAnimationEqualPSNR(
     WebPData* const webp_data) {
-  // Rearrange frames.
-  std::sort(frames_.begin(), frames_.end(),
-            [](const FrameData& A, const FrameData& B) -> bool {
-              return A.timestamp_ms < B.timestamp_ms;
-            });
+  // GenerateAnimation failed.
+  if (config_.quality == -1) return kByteBudgetError;
```

Comment:
You assume that GenerateAnimation() has been called before GenerateAnimationEqualPSNR(), which is a confusing API. Can't you just call GenerateAnimation() yourself here?
Same thing for SortFrames() which could be called automatically.